numpy/lib/function_base.py

```diff
@@ -327,6 +327,39 @@ def _hist_bin_scott(x):
     return (24.0 * np.pi**0.5 / x.size)**(1.0 / 3.0) * np.std(x)
 
 
+def _hist_bin_ise(x):
+    """
+    Histogram bin estimator based on minimizing the estimated integrated squared error (ISE).
+
+    The number of bins is chosen by minimizing the estimated ISE against the unknown true distribution.
+    The ISE is estimated using cross-validation and can be regarded as a generalization of Scott's rule.
+    https://en.wikipedia.org/wiki/Histogram#Scott.27s_normal_reference_rule
+
+    Parameters
+    ----------
+    x : array_like
+        Input data that is to be histogrammed, trimmed to range. May not
+        be empty.
+
+    Returns
+    -------
+    h : An estimate of the optimal bin width for the given data.
+    """
+
+    n = x.size
+    ptp_x = np.ptp(x)
+    if n <= 1 or ptp_x == 0:
+        return ptp_x
+
+    def jhat(nbins):
+        hh = ptp_x / nbins
+        N_k = np.histogram(x, bins=nbins)[0] / n
+        return (2 - (n + 1) * N_k.dot(N_k)) / (n - 1) / hh
+
+    nbins = min(np.arange(1, max(n // 4, 2)), key=jhat)
```

Comment:
It seems like the upper bound should scale sub-linearly, given that that is the case for all the other estimators.

I also wonder if it would make more sense to use one of the other (ideally more conservative) bin estimators for the upper bound.

Given these concerns, perhaps the square root rule would make sense, given the general arguments that the number of bins should scale like the cube root? Maybe `max(100, int(np.sqrt(n)))` to avoid strange behavior for small `n`?